stdlib/public/core/ClosedRange.swift

```diff
@@ -47,26 +47,64 @@ public struct ClosedRangeIndex<Bound> : Comparable
     }
   }
   
-  internal func _predecessor(upperBound limit: Bound) -> ClosedRangeIndex {
+  internal func _predecessor(
+    lowerBound: Bound, upperBound limit: Bound
+  ) -> ClosedRangeIndex {
     switch _value {
-    case .inRange(let x): return ClosedRangeIndex(x.advanced(by: -1))
-    case .pastEnd: return ClosedRangeIndex(limit)
+    case .inRange(let x):
+      _precondition(x > lowerBound, "Incrementing past start index")
+      return ClosedRangeIndex(x.advanced(by: -1))
+    case .pastEnd: 
+      _precondition(limit >= lowerBound, "Incrementing past start index")
+      return ClosedRangeIndex(limit)
     }
   }
   
   internal func _advanced(
-    by n: Bound.Stride, upperBound limit: Bound
+    by n: Bound.Stride, lowerBound: Bound, upperBound limit: Bound
   ) -> ClosedRangeIndex {
     switch _value {
     case .inRange(let x):
       let d = x.distance(to: limit)
-      if n <= d { return ClosedRangeIndex(x.advanced(by: n)) }
+      if n <= d {
+        let newPosition = x.advanced(by: n)
+        _precondition(newPosition >= lowerBound,
+          "Advancing past start index")
+        return ClosedRangeIndex(newPosition)
+      }
       if d - -1 == n { return ClosedRangeIndex() }
       _preconditionFailure("Advancing past end index")
     case .pastEnd:
-      return n == 0
-      ? self :
-      ClosedRangeIndex(limit)._advanced(by: n, upperBound: limit)
+      if n == 0 {
+        return self
+      } else if n > 0 {
+        _preconditionFailure("Advancing past end index")
+      } else {
+        return ClosedRangeIndex(limit)._advanced(
+          by: (n + 1),
+          lowerBound: lowerBound,
+          upperBound: limit
+        )
+      }
+    }
+  }
+
+  internal func _distance(
+    to: ClosedRangeIndex<Bound>, upperBound limit: Bound
+  ) -> Bound.Stride {
+    switch (_value, to._value) {
+    case let (.inRange(left), .inRange(right)):
+      // in range <--> in range
+      return left.distance(to: right)
+    case let (.inRange(left), .pastEnd):
+      // in range --> end
+      return 1 + left.distance(to: limit)
+    case let (.pastEnd, .inRange(right)):
+      // in range <-- end
+      return 1 + limit.distance(to: right)
```

Comment:
Yes please. :+1:

Also... Since we're here already, it might make sense to change the requirement to `Bound.Stride : SignedInteger` to reflect this.